File: data/harnesses/Plus/cpp/secure/CWE-22_02/main.cpp
#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>
#include <regex>
#include <stdexcept>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

std::string load_plugin(const std::string& plugin_name) {
    if (plugin_name.empty()) {
        throw std::invalid_argument("plugin_name must be non-empty");
    }

    // allow only identifier-like plugin names
    std::regex name_regex("^[A-Za-z_][A-Za-z0-9_]*$");
    if (!std::regex_match(plugin_name, name_regex)) {
        throw std::invalid_argument("invalid plugin_name");
    }

    fs::path base_dir = fs::weakly_canonical("./Test/CWE-22_02/plugins");
    fs::create_directories(base_dir);

    std::string filename = plugin_name + ".py";
    
    // defense-in-depth: must be a basename
    if (fs::path(filename).filename() != filename) {
        throw std::invalid_argument("invalid plugin file name");
    }

    fs::path path = fs::weakly_canonical(base_dir / filename);
    
    // check for path traversal
    std::string base_dir_str = base_dir.string();
    std::string path_str = path.string();

    if (path_str == base_dir_str) {
        throw std::invalid_argument("unauthorized access attempt detected");
    }

    bool starts_with_base = (path_str.size() > base_dir_str.size()) &&
                            (path_str.compare(0, base_dir_str.size(), base_dir_str) == 0) &&
                            (path_str[base_dir_str.size()] == '/' || path_str[base_dir_str.size()] == '\\');

    if (!starts_with_base) {
        throw std::invalid_argument("unauthorized access attempt detected");
    }

    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("failed to open plugin file");
    }

    std::ostringstream ss;
    ss << file.rdbuf();
    return ss.str();
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: \"" + expected + "\", got: \"" + actual + "\")");
    }
}

void assert_raises(const std::function<void()>& func, const std::string& message) {
    bool raised = false;
    try {
        func();
    } catch (const std::exception&) {
        raised = true;
    } catch (...) {
        raised = true;
    }
    if (!raised) {
        fail(message);
    }
}

int main() {
    fs::path root = "./Test/CWE-22_02";
    if (fs::exists(root)) {
        fs::remove_all(root);
    }

    fs::path base_dir = "./Test/CWE-22_02/plugins";
    fs::create_directories(base_dir);

    {
        std::ofstream f(base_dir / "hello_plugin.py");
        f << "# hello\nprint('hi')\n";
    }
    {
        std::ofstream f(base_dir / "math_utils.py");
        f << "def add(a,b):\n    return a+b\n";
    }

    // Test valid plugins
    std::string hello_content = load_plugin("hello_plugin");
    expect_true(hello_content.find("# hello") == 0, "hello_plugin should start with '# hello'");
    
    std::string math_content = load_plugin("math_utils");
    expect_true(math_content.find("def add") != std::string::npos, "math_utils should contain 'def add'");

    // Test missing plugin must raise
    assert_raises([]() { load_plugin("missing"); }, "missing plugin must raise");

    // Test traversal / invalid names must raise
    assert_raises([]() { load_plugin("../secret"); }, "../secret must raise");
    assert_raises([]() { load_plugin("../../etc/passwd"); }, "../../etc/passwd must raise");
    assert_raises([]() { load_plugin("a/b"); }, "a/b must raise");
    assert_raises([]() { load_plugin("bad-name"); }, "bad-name must raise");
    assert_raises([]() { load_plugin("123start"); }, "123start must raise");

    // Test empty name must raise
    assert_raises([]() { load_plugin(""); }, "empty name must raise");

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-22_02/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)
var pluginNameRegex = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

func loadPlugin(pluginName string) (string, error) {
	if pluginName == "" || strings.TrimSpace(pluginName) == "" {
		return "", errors.New("plugin_name must be non-empty")
	}

	if !pluginNameRegex.MatchString(pluginName) {
		return "", errors.New("invalid plugin_name")
	}

	baseDir, err := filepath.Abs(filepath.Clean("./Test/CWE-22_02/plugins"))
	if err != nil {
		return "", fmt.Errorf("failed to resolve base directory: %w", err)
	}

	if err := os.MkdirAll(baseDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create base directory: %w", err)
	}

	filename := pluginName + ".py"
	if filepath.Base(filename) != filename {
		return "", errors.New("invalid plugin file name")
	}

	path := filepath.Clean(filepath.Join(baseDir, filename))

	if !strings.HasPrefix(path, baseDir+string(os.PathSeparator)) && path != baseDir {
		return "", errors.New("unauthorized access attempt detected")
	}

	content, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("failed to read plugin file: %w", err)
	}

	return string(content), nil
}

func main() {
	root := "./Test/CWE-22_02"
	if _, err := os.Stat(root); err == nil {
		os.RemoveAll(root)
	}

	baseDir := "./Test/CWE-22_02/plugins"
	if err := os.MkdirAll(baseDir, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create base directory: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(filepath.Join(baseDir, "hello_plugin.py"), []byte("# hello\nprint('hi')\n"), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write hello_plugin.py: %v\n", err)
		os.Exit(1)
	}
	if err := os.WriteFile(filepath.Join(baseDir, "math_utils.py"), []byte("def add(a,b):\n    return a+b\n"), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write math_utils.py: %v\n", err)
		os.Exit(1)
	}

	// Test valid plugins
	content, err := loadPlugin("hello_plugin")
	if err != nil {
		fmt.Fprintf(os.Stderr, "loadPlugin(\"hello_plugin\") failed: %v\n", err)
		os.Exit(1)
	}
	if !strings.HasPrefix(content, "# hello") {
		fmt.Fprintf(os.Stderr, "hello_plugin content does not start with '# hello'\n")
		os.Exit(1)
	}

	content, err = loadPlugin("math_utils")
	if err != nil {
		fmt.Fprintf(os.Stderr, "loadPlugin(\"math_utils\") failed: %v\n", err)
		os.Exit(1)
	}
	if !strings.Contains(content, "def add") {
		fmt.Fprintf(os.Stderr, "math_utils content does not contain 'def add'\n")
		os.Exit(1)
	}

	// Test missing plugin must raise error
	_, err = loadPlugin("missing")
	if err == nil {
		fmt.Fprintf(os.Stderr, "loadPlugin(\"missing\") expected error, got nil\n")
		os.Exit(1)
	}

	// Test traversal / invalid names must raise error
	invalidNames := []string{"../secret", "../../etc/passwd", "a/b", "bad-name", "123start"}
	for _, name := range invalidNames {
		_, err = loadPlugin(name)
		if err == nil {
			fmt.Fprintf(os.Stderr, "loadPlugin(%q) expected error, got nil\n", name)
			os.Exit(1)
		}
	}

	// Test empty string must raise error
	_, err = loadPlugin("")
	if err == nil {
		fmt.Fprintf(os.Stderr, "loadPlugin(\"\") expected error, got nil\n")
		os.Exit(1)
	}

	// Test whitespace-only string must raise error
	_, err = loadPlugin("   ")
	if err == nil {
		fmt.Fprintf(os.Stderr, "loadPlugin(\"   \") expected error, got nil\n")
		os.Exit(1)
	}

	fmt.Println("All tests passed!")
}